### PR TITLE
Solve square problem to initialize regression problem using objective_at_theta() in parmest

### PR DESCRIPTION
**Pull request by kanishka-ghosh**
## Fixes # .
Implements changes proposed in issue #2377 

## Summary/Motivation:
Add code and tests to solve a square problem instance in order to initialize the regression problem in parmest

## Changes proposed in this PR:
- Solve square problem instances of the model for each scenario using parameter values provided by user or values used to initialize fitted parameters; create extensive form of the model before calling `theta_est()` for parameter estimation.
  - Call `objective_at_theta(initialize_parmest_model=True)` which calls `_Q_at_theta()`
  - In `_Q_at_theta()`, iterate over scenarios, create and solve instance of Pyomo model, and store solved model instance in scenario dictionary
  - Call function `_create_EF_from_scen_dict()` to create the extensive form model with the scenario dictionary from above and set flag `model_initialized=True` 
  - Call `theta_est()` to solve parameter estimation problem using extensive form of model from above (skip building the extensive form using `theta_est()`).
- Create instance of Rooney-Biegler problem with `Constraints` to test square solve initialization
- Add tests for square solve functionality

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by jsiirola** [comment]
@kanishka-ghosh Note that the failing test is due to code coverage of your patch is slightly below the average coverage of the Pyomo codebase.

**Comment by blnicho** [comment]
@kanishka-ghosh I merged PR #2352 from Kate which caused some conflicts with this PR. Could you try to resolve the conflicts and let me know if you have any issues?

**Comment by kanishka-ghosh** [comment]
@blnicho I resolved the merge conflicts

**Comment by blnicho** [comment]
@kaklise could you review this PR?

File: pyomo/contrib/parmest/parmest.py [hunk 1]
```diff
@@ -848,99 +910,117 @@ def leaveNout_bootstrap_test(self, lNo, lNo_samples, bootstrap_samples,
...
-        theta_names = theta_values.columns
-        all_thetas = theta_values.to_dict('records')
+        if theta_values is None:
+            all_thetas = {} # dictionary to store fitted variables
+            theta_names = self.theta_names
+        else:
+            assert isinstance(theta_values, pd.DataFrame)
+
+            # for parallel code we need to use lists and dicts in the loop
+            theta_names = theta_values.columns
```
Comment (kaklise) [inline pyomo/contrib/parmest/parmest.py:974]: Is it ok if `theta_names` does not equal `self.theta_names`?

File: pyomo/contrib/parmest/tests/test_parmest.py [hunk 2]
```diff
@@ -581,55 +581,109 @@ def total_cost_rule(model):
...
+                'tol': 1e-8,
+                }
+
+        self.data = data
+        self.pest = parmest.Estimator(rooney_biegler_model_with_constraint, data, theta_names, SSE,
+                solver_options=solver_options,tee=True)
+
+    def test_theta_est_with_square_initialization(self):
+        obj_init = self.pest.objective_at_theta(initialize_parmest_model=True)
+        print("obj_init = ", obj_init)
```
Comment (kaklise) [inline pyomo/contrib/parmest/tests/test_parmest.py:661]: Remove print statement

File: pyomo/contrib/parmest/tests/test_parmest.py [hunk 3]
```diff
@@ -581,55 +581,109 @@ def total_cost_rule(model):
...
+        objval, thetavals = self.pest.theta_est()
+
+        self.assertAlmostEqual(objval, 4.3317112, places=2)
+        self.assertAlmostEqual(thetavals['asymptote'], 19.1426, places=2)  # 19.1426 from the paper
+        self.assertAlmostEqual(thetavals['rate_constant'], 0.5311, places=2)  # 0.5311 from the paper
+
+    def test_theta_est_with_square_initialization_diagnostic_mode_true(self):
+        self.pest.diagnostic_mode = True
+        obj_init = self.pest.objective_at_theta(initialize_parmest_model=True)
+        print("obj_init = ", obj_init)
```
Comment (kaklise) [inline pyomo/contrib/parmest/tests/test_parmest.py:680]: Remove print statement

File: pyomo/contrib/parmest/parmest.py [hunk 4]
```diff
@@ -848,99 +910,117 @@ def leaveNout_bootstrap_test(self, lNo, lNo_samples, bootstrap_samples,
...
-        for Theta in local_thetas:
-            obj, thetvals, worststatus = self._Q_at_theta(Theta)
+        if len(all_thetas) > 0:
+            for Theta in local_thetas:
+                obj, thetvals, worststatus = self._Q_at_theta(Theta, initialize_parmest_model=initialize_parmest_model)
+                if worststatus != pyo.TerminationCondition.infeasible:
+                     all_obj.append(list(Theta.values()) + [obj])
+                # DLW, Aug2018: should we also store the worst solver status?
+        else:
+            obj, thetvals, worststatus = self._Q_at_theta(thetavals={}, initialize_parmest_model=initialize_parmest_model)
```
Comment (kaklise) [inline pyomo/contrib/parmest/parmest.py:990]: When `theta_values` is None, the fixed values are used, and the returned `obj_at_theta` is an empty dataframe.  Can `obj_at_theta` return the fixed theta values and objective value even when `theta_values` is None?

File: pyomo/contrib/parmest/parmest.py [hunk 5]
```diff
@@ -524,31 +526,35 @@ def _Q_opt(self, ThetaVals=None, solver="ef_ipopt",
...
         Return the objective function value with fixed theta values.
-        
+
         Parameters
         ----------
         thetavals: dict
             A dictionary of theta values.
+        
+        initialize_parmest_model: boolean
+            True: Solve square problem instance, build extensive form of the model for
```
[the remark this answers is not in the document]
Comment (kanishka-ghosh) [reply pyomo/contrib/parmest/parmest.py:556]: Default value is `False`.  Replaced line 556 to begin with `If True:`

**Review by blnicho — APPROVED** [review]

**Comment by blnicho** [comment]
@kaklise could you review this PR again when you have a chance?

**Comment by kanishka-ghosh** [comment]
@adowling2 volunteered to review

File: pyomo/contrib/parmest/parmest.py [hunk 6]
```diff
@@ -607,21 +649,56 @@ def _Q_at_theta(self, thetavals):
...
             
         retval = totobj / len(senario_numbers) # -1??
+        if initialize_parmest_model:
+            # create extensive form of the model using scenario dictionary
+            if len(scen_dict) > 0:
+                for scen in scen_dict.values():
+                    scen._mpisppy_probability = 1 / len(scen_dict)
+            
+            if use_mpisppy:
+                ef = sputils._create_EF_from_scen_dict(scen_dict,
```
Comment (adowling2) [inline pyomo/contrib/parmest/parmest.py:676]: This line is not covered with testing, which makes me suspect the tests are not considering `initialize_parmest_mode=True` and `mpi-sppy`. Does the testing harness have `mpi-sppy` installed?

**Review by adowling2 — COMMENTED** [review]
Looks good. One question about the testing suite.

File: pyomo/contrib/parmest/parmest.py (hunk 6 above)
Comment (blnicho) [reply pyomo/contrib/parmest/parmest.py:676]: I don't think our testing harness has mpi-sppy installed

**Review by adowling2 — APPROVED** [review]

File: pyomo/contrib/parmest/parmest.py (hunk 6 above)
Comment (adowling2) [reply pyomo/contrib/parmest/parmest.py:676]: I just approved the PR then. No `mpi-sppy` in the testing harness explains the code coverage.

File: pyomo/contrib/parmest/parmest.py (hunk 1 above)
Comment (kaklise) [reply pyomo/contrib/parmest/parmest.py:988]: Given the way parmest modifies the model, I think these two list of names needs to match.  A check would be good.

File: pyomo/contrib/parmest/parmest.py [hunk 7]
```diff
@@ -607,21 +649,41 @@ def _Q_at_theta(self, thetavals):
...
-
+        if initialize_parmest_model:
+            # create extensive form of the model with using scenario dictionary
+            if len(scen_dict) > 0:
+                for scen in scen_dict.values():
+                    scen._mpisppy_probability = 1 / len(scen_dict)
+            EF_instance = local_ef._create_EF_from_scen_dict(scen_dict,
+                                                    EF_name="_Q_at_theta",
+                                                    nonant_for_fixed_vars=True)
+
```
[the remark this answers is not in the document]
Comment (kaklise) [reply pyomo/contrib/parmest/parmest.py:683]: I'm still running into some issues when using mpisppy.  I updated that package to the latest version.  When `use_mpisppy` is False, the tests pass.  Then `use_mpisppy` is True, I had to change the following to get the tests to pass:
- Comment out Line168 (`scen_name_list=None, # Deprecated?`) 
- Comment out Line 678 (`suppress_warnings=True`)
- Change `ef` in line 676 to `EF_instance`

[the remark this answers is not in the document]
Comment (kanishka-ghosh) [reply pyomo/contrib/parmest/parmest.py:684]: @kaklise Thank you for pointing these out.  For line 168, I commented out `scen_name_list=None, # Deprecated?` when `use_mpisppy = True`.  However, when `use_mpisppy = False`, `instance._mpisppy_mode_list` is created using `scenario_tree` imported from `pyomo.contrib.parmest.utils` where `scen_name_list` is a required argument.

Therefore, I'm using `scen_name_list` argument when `use_mpisppy=False` and removed the argument when `use_mpisppy=True` through an `if-else` statement

File: pyomo/contrib/parmest/parmest.py (hunk 1 above)
Comment (kanishka-ghosh) [reply pyomo/contrib/parmest/parmest.py:998]: I added a check for `theta_names` to the equal to `self.theta_names` only when `initialize_parmest_model=True`. I am not sure how to check this for the different model variants being tested in lines 291 - 418 of https://github.com/Pyomo/pyomo/blob/main/pyomo/contrib/parmest/tests/test_parmest.py

Comment (kanishka-ghosh) [reply pyomo/contrib/parmest/parmest.py:998]: @kaklise I ran into a few issues while writing the check to compare `theta_names` and `self.theta_names`: the check fails for certain model variants specified in the test in lines 291 - 412 of https://github.com/Pyomo/pyomo/blob/main/pyomo/contrib/parmest/tests/test_parmest.py: 

1. Check fails for model variants `param_index` and `vars_index` since `theta_names` and column names of `theta_vals` are different
2. Check fails for model variants `vars_quoted_index` and `vars_str_index`: `theta_est()` modifies `self.theta_names` in line 368 of https://github.com/Pyomo/pyomo/blob/main/pyomo/contrib/parmest/parmest.py and removes quotes from indexed parameter names (e.g. `theta['asymptote']` becomes `theta[asymptote]`)

I am working on fixing these check fails.  Please let me know if you have any suggestions/comments

**Comment by blnicho** [comment]
@kanishka-ghosh I think we should wait on this one for the release. I was taking a closer look at the test failures before your commits today and I'm not convinced that your fix is what we want. In fact, I think the tests you added for the functionality introduced in this PR are missing some important edge cases. For example, when I try adding the `initialize_parmest_model=True` flag to the `test_parmest_basics` test (https://github.com/kanishka-ghosh/pyomo/blob/square_solve_obj_at_theta/pyomo/contrib/parmest/tests/test_parmest.py#L411) it fails with a `ZeroDivisionError`. `test_parmest_basics` tests different configurations/structures for the same model so it's concerning to me that it doesn't work. If this is an invalid test case for your `initialize_parmest_model` flag let me know.

**Review by blnicho — APPROVED** [review]
Two non-functional edits to the documentation and I think this is ready.

**Comment by blnicho** [comment]
@kaklise could you take another look at this PR?

File: doc/OnlineDocs/contributed_packages/parmest/driver.rst [hunk 8]
```diff
@@ -156,3 +156,16 @@ expression which is used to define "SecondStageCost".  The objective
...
+--------------------------------------------------------------------
+
+To check the quality of initial guess values provided for the fitted parameters, we suggest solving a 
+square instance of the problem prior to solving the parameter estimation problem using the following steps:
+
+1. Create :class:`~pyomo.contrib.parmest.parmest.Estimator` object
+
+2. Call :class:`~pyomo.contrib.parmest.parmest.Estimator.objective_at_theta` with optional argument ``(initialize_parmest_model=True)``. Different initial guess values for the fitted parameters can be provided using optional argument `theta_values` (**Pandas Dataframe**)
+
+3. (Optional) Initialize parameter estimation solve from the square problem solution by setting IPOPT solver option ``bound_push`` to a value of ``1e-8``
```
Comment (kaklise) [inline doc/OnlineDocs/contributed_packages/parmest/driver.rst:169]: This step would need to be done when the Estimator object is created.  I suggest moving this step into step 1 and formatting the optional input as a dictionary `solver_options={'bound_push': 1e-8}`

File: doc/OnlineDocs/contributed_packages/parmest/driver.rst [hunk 9]
```diff
@@ -156,3 +156,14 @@ expression which is used to define "SecondStageCost".  The objective
...
+--------------------------------------------------------------------
+
+To check the quality of initial guess values provided for the fitted parameters, we suggest solving a 
+square instance of the problem prior to solving the parameter estimation problem using the following steps:
+
+1. Create :class:`~pyomo.contrib.parmest.parmest.Estimator` object. To initialize the parameter estimation solve from the square problem solution, set optional argument ``solver_options = {bound_push: 1e-8}``.
+
+2. Call :class:`~pyomo.contrib.parmest.parmest.Estimator.objective_at_theta` with optional argument ``(initialize_parmest_model=True)``. Different initial guess values for the fitted parameters can be provided using optional argument `theta_values` (**Pandas Dataframe**)
+
+4. Solve parameter estimation problem by calling :class:`~pyomo.contrib.parmest.parmest.Estimator.theta_est`
```
[the remark this answers is not in the document]
Comment (kanishka-ghosh) [reply doc/OnlineDocs/contributed_packages/parmest/driver.rst]: Sorry, I missed this.

**Review by kaklise — APPROVED** [review]
